pkg/ruler/frontend_decoder.go

```diff
@@ -86,3 +98,73 @@ func (j JsonDecoder) scalarToPromQLVector(scalar model.Scalar) promql.Vector {
 		Metric: labels.Labels{},
 	}}
 }
+
+func (p ProtobufDecoder) ContentType() string {
+	return tripperware.QueryResponseCortexMIMEType
+}
+
+func (p ProtobufDecoder) Decode(body []byte) (promql.Vector, Warning, error) {
+	resp := tripperware.PrometheusResponse{}
+	if err := resp.Unmarshal(body); err != nil {
+		return nil, nil, err
+	}
+
+	if resp.Status == statusError {
+		return nil, resp.Warnings, fmt.Errorf("failed to execute query with error: %s", resp.Error)
+	}
+
+	switch resp.Data.ResultType {
+	case "scalar":
+		data := struct {
+			Type   model.ValueType `json:"resultType"`
+			Result json.RawMessage `json:"result"`
+		}{}
+
+		if err := json.Unmarshal(resp.Data.Result.GetRawBytes(), &data); err != nil {
+			return nil, nil, err
+		}
+
+		var s model.Scalar
+		if err := json.Unmarshal(data.Result, &s); err != nil {
+			return nil, nil, err
+		}
+		return p.scalarToPromQLVector(s), resp.Warnings, nil
+	case "vector":
+		return p.vectorToPromQLVector(resp.Data.Result.GetVector()), resp.Warnings, nil
+	default:
+		return nil, resp.Warnings, errors.New("rule result is not a vector or scalar")
+	}
+}
+
+func (p ProtobufDecoder) vectorToPromQLVector(vector *tripperware.Vector) promql.Vector {
+	v := make([]promql.Sample, 0, len(vector.Samples))
+	for _, sample := range vector.Samples {
+		metric := cortexpb.FromLabelAdaptersToLabels(sample.Labels)
+
+		if sample.Sample != nil {
+			v = append(v, promql.Sample{
+				T:      sample.Sample.TimestampMs,
+				F:      sample.Sample.Value,
+				Metric: metric,
+			})
+		}
+
+		if sample.RawHistogram != nil {
+			v = append(v, promql.Sample{
+				T:      sample.RawHistogram.TimestampMs,
+				H:      cortexpb.FloatHistogramProtoToFloatHistogram(*sample.RawHistogram),
+				Metric: metric,
+			})
+		}
+	}
+
+	return v
+}
+
+func (p ProtobufDecoder) scalarToPromQLVector(s model.Scalar) promql.Vector {
```

Comment:
Nit: this doesn't need to be a method of `ProtobufDecoder`.

Reply:
Can we reuse the same function from `jsonDecoder`?

Reply:
sure, thanks for catching.